Home page component

ryanlieu opened the pull request:
### Summary of Changes

Slightly larger PR that encapsulates the changes needed to create a separate home page component. Work is roughly split into three parts:

- Making SearchBar a "smart" (container) component vs. a "dumb" (presentational) component so that it can be re-used on multiple pages/locations (for example in case we want to put the search bar into the nav bar) 
- Putting SearchBar, Bookmarks, and PopularTables components into a new Homepage component that can be reached by navigating to /
- Fixing the breadcrumb component on TableDetails page to intelligently route back to search results (if table was accessed through search) or home (if table was accessed by directly loading the page or navigating from bookmarks/popular tables). 

### Tests

Will update tests after implementation gets looked over.

### Documentation

_What documentation did you add or modify and why? Add any relevant links then remove this line_

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes, including screenshots of any UI changes. 
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public python functions and the classes in the PR contain docstrings that explain what it does
- [ ] PR passes all tests documented in the [developer guide](https://github.com/lyft/amundsenfrontendlibrary/blob/master/docs/developer_guide.md#testing)
- [ ] I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"

ryanlieu commented:
@ttannis @danwom this PR is good to review for implementation. Will write/update tests in parallel

ttannis commented on amundsen_application/static/js/components/HomePage/index.tsx:
```diff
@@ -0,0 +1,67 @@
+import * as React from 'react';
```
nit: follow a loose import hierarchy of 3rd party imports (e.g. react/redux) --> absolute imports (e.g. components/ducks) --> relative imports (e.g. ../someFile)

ttannis commented on amundsen_application/static/js/components/HomePage/index.tsx:
```diff
@@ -0,0 +1,67 @@
+    super(props);
+  }
+
+  createSearchOptions = (pageIndex: number, selectedTab: ResourceType) => {
```
Why is this method here?

ttannis commented on amundsen_application/static/js/components/HomePage/index.tsx:
```diff
@@ -0,0 +1,67 @@
+    this.props.searchReset();
+  }
+
+  onSearchBarSubmit = (searchTerm: string): void => {
```
Why is this method here?

ttannis commented on amundsen_application/static/js/components/HomePage/index.tsx:
```diff
@@ -0,0 +1,67 @@
+          <div className="row">
+            <div className="col-xs-12 col-md-offset-1 col-md-10">
+              <div className="search-list-container">
+                <SearchBar />
```
`SearchBar` should be outside of the `.search-list-container` element. `BookMarkList` & `PopularTables` should also be in their own `<div>` as well with whatever classname we want to apply to create the appropriate spacing in between them as requested by Design. See next comment in `styles.css`.

ttannis commented on amundsen_application/static/js/components/HomePage/styles.scss:
```diff
@@ -0,0 +1,13 @@
+@import 'variables';
+
+.home-page {
+  .search-list-container {
```
This style should have it's own name. In addition we have to increase the spacing between each of the lists on this page as requested by Design.

ttannis commented on amundsen_application/static/js/components/SearchPage/SearchBar/index.tsx:
```diff
@@ -11,11 +11,16 @@ import {
   SYNTAX_ERROR_PREFIX,
   SYNTAX_ERROR_SPACING_SUFFIX,
 } from './constants';
+import { RouteComponentProps, withRouter } from 'react-router';
```
nit: Same comment about import order.

ttannis commented on amundsen_application/static/js/components/SearchPage/SearchBar/index.tsx:
```diff
@@ -11,11 +11,16 @@ import {
+  searchTerm: string;
+}
 
 export interface SearchBarProps {
```
You can now call this `OwnProps` and declare:
`export type SearchBarProps = StateFromProps & OwnProps & RouteComponentProps<any>;`
and then be able to keep:
 `export class SearchBar extends React.Component<SearchBarProps, SearchBarState>`

ttannis commented on amundsen_application/static/js/components/SearchPage/index.tsx:
```diff
@@ -258,7 +243,7 @@ export class SearchPage extends React.Component<SearchPageProps, SearchPageState
         <div className="row">
           <div className="col-xs-12 col-md-offset-1 col-md-10">
-            <SearchBar handleValueSubmit={ this.onSearchBarSubmit } searchTerm={ searchTerm }/>
+            <SearchBar />
```
We can also remove `onSearchBarSubmit` from this file as well as it's test.

ttannis commented on amundsen_application/static/js/components/TableDetail/index.tsx:
```diff
@@ -126,6 +127,16 @@ export class TableDetail extends React.Component<TableDetailProps & RouteCompone
     })
   };
 
+  getBreadcrumb() {
```
nit: `renderBreadcrumb` would be a more appropriate name.

ttannis commented on amundsen_application/static/js/components/TableDetail/index.tsx:
```diff
@@ -330,7 +341,7 @@ export class TableDetail extends React.Component<TableDetailProps & RouteCompone
         innerContent = (
           <div className="container table-detail">
-            <Breadcrumb path='/' text='Search Results'/>
+            { this.getBreadcrumb() }
```
I'm fine with this approach for now, if we merge it in as is we should file a ticket to shift to the the alternative of putting this logic in `Breadcrumb` since we'll have to use it on the `ProfilePage` as well. cc @danwom since you're working on getting the `ProfilePage` ready.

ttannis commented on amundsen_application/static/js/components/TableDetail/index.tsx:
```diff
@@ -376,6 +387,7 @@ export const mapStateToProps = (state: GlobalState) => {
     isLoading: state.tableMetadata.isLoading,
     statusCode: state.tableMetadata.statusCode,
     tableData: state.tableMetadata.tableData,
+    searchTerm: state.search.search_term,
```
nit: maintain alphabetical order for object keys.

ttannis commented on amundsen_application/static/js/index.tsx:
```diff
@@ -17,7 +17,8 @@ import NavBar from './components/NavBar';
 import ProfilePage from './components/ProfilePage';
-import SearchPage from './components/SearchPage';
+import SearchPage from './components/SearchPage'; 
+import HomePage from './components/HomePage'
```
nit: place `HomePage` higher up in alphabetical order, and looks like I missed this file when we switched to absolute imports. We can remove the relative imports in this file and switch to:
`import Foo from 'components/Foo'` and `import Thing from 'ducks/pathToThing'`

ryanlieu commented:
@ttannis this PR should be good to look over again, I've addressed your comments in the subsequent commits. 

One quick thing I was wondering about was the hard coding of the resource type in some of the searches. With the changes to search bar and breadcrumb, redirects are hard coded to selectedTab=table as there is no global redux store for selectedTab and page. This is fine for now as there's no other tabs we're currently displaying. I think it'd make sense to file a follow up task to add those two search variables to the global store a la searchTerm, but lmk if I should address that in this PR.

ttannis commented on amundsen_application/static/js/components/HomePage/styles.scss:
```diff
@@ -0,0 +1,13 @@
+@import 'variables';
+
+.home-page {
```
Would be curious to see a screenshot of what this looks like -- is the first `SearchBar` pushed further down on this page than it is on the `SearchPage`? Is the `BookmarkList` further away fro the `SearchBar` on this page as well?

ryanlieu replied:
oh whoops shouldn't be applying the home-element-container class to SearchBar, will attach a screenshot of the fixed version

ryanlieu replied:
<img width="1294" alt="Screen Shot 2019-06-11 at 2 21 45 PM" src="https://user-images.githubusercontent.com/8528315/59307866-48eace00-8c54-11e9-8e2d-a0413c375114.png">

ttannis commented:
> @ttannis this PR should be good to look over again, I've addressed your comments in the subsequent commits.
> 
> One quick thing I was wondering about was the hard coding of the resource type in some of the searches. With the changes to search bar and breadcrumb, redirects are hard coded to selectedTab=table as there is no global redux store for selectedTab and page. This is fine for now as there's no other tabs we're currently displaying. I think it'd make sense to file a follow up task to add those two search variables to the global store a la searchTerm, but lmk if I should address that in this PR.

That was an intentional decision for the default selected tab to be table, and the default pageIndex to be 0.

ryanlieu commented on amundsen_application/static/js/components/common/Breadcrumb/index.tsx:
```diff
@@ -1,14 +1,23 @@
+  let path = '/';
+  let text = 'Home';
+  if (props.searchTerm) {
+    path = `/search?searchTerm=${props.searchTerm}&selectedTab=table&pageIndex=0`
```
able to fix this with a one-liner, will include in this PR. Just doing a check on componentDidMount to see if term is the same as the searchTerm in redux. If it is we can just not call searchAll as the results for that searchTerm are stored in redux already.

ryanlieu commented:
@ttannis should be good to go for a final implementation pass, working on unit tests atm

danwom commented on amundsen_application/static/js/components/HomePage/index.tsx:
```diff
@@ -0,0 +1,56 @@
+
+  render() {
+    return (
+      <>
```
The wrapper `<>` element is useful when you have multiple "root" elements on a JSX node. For example
```
// Not allowed
(
  <h1> header </h1>
  <body> body</body>
)

// Allowed
(
  <>
    <h1> header </h1>
    <body> body</body>
  </>
)


```

ttannis replied:
i.e. you don't need the `<>` here

danwom commented on amundsen_application/static/js/components/common/Breadcrumb/index.tsx:
```diff
@@ -1,14 +1,28 @@
+  let path = props.path;
+  let text = props.text;
+  if (props.searchTerm) {
+    path = `/search?searchTerm=${props.searchTerm}&selectedTab=table&pageIndex=0`
```
Also worth considering: You could do `path = 'search?searchTerm=${props.searchTerm}'` and leave `selectedTab` and `pageIndex` as empty. Then the `SearchPage` could use a default value for these.

ryanlieu replied:
the only issue with this is that the back button then requires two clicks to go back as the selectedTab and pageIndex are injected into the url. Will address this in a follow up PR.

ttannis replied:
^ not sure I understand this, but can we chat about it on slack or in person

ttannis commented on amundsen_application/static/js/components/HomePage/tests/index.spec.tsx:
```diff
@@ -0,0 +1,73 @@
+import BookmarkList from 'components/common/Bookmark/BookmarkList';
+import PopularTables from 'components/common/PopularTables';
+
+describe('SearchBar', () => {
```
change title to 'HomePage'

ttannis commented on amundsen_application/static/js/components/HomePage/tests/index.spec.tsx:
```diff
@@ -0,0 +1,73 @@
+  const setup = (propOverrides?: Partial<HomePageProps>) => {
+    const props: HomePageProps = {
+      searchReset: jest.fn(),
+      history: {
```
worthy mention: I'll see if i have time to look into how to better write tests that leverage the router. This works but I recognize it's pretty inconvenient.

ttannis commented on amundsen_application/static/js/components/HomePage/tests/index.spec.tsx:
```diff
@@ -0,0 +1,78 @@
+
+  describe('render', () => {
+    it('contains Searchbar, BookmarkList, and PopularTables', () => {
+      wrapper.instance().render();
```
Doesn't this pass without needing to call `render()`?

ttannis commented on amundsen_application/static/js/components/ProfilePage/index.tsx:
```diff
@@ -86,7 +86,7 @@ export class ProfilePage extends React.Component<ProfilePageProps> {
           <div className="row">
             <div className="col-xs-12 col-md-offset-1 col-md-10">
-              <Breadcrumb path='/' text='Search Results'/>
+              <Breadcrumb path="/" text="Home" />
```
This one was fine to leave as `<Breadcrumb />`, but since we haven't exposed this in search yet it makes sense to leave it as directing to `Home`. Can you add a comment about removing the hardcoded path to `Home` when we are ready to expose this in production?

ttannis commented on amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx:
```diff
@@ -680,10 +645,7 @@ describe('SearchPage', () => {
-        handleValueSubmit: wrapper.instance().onSearchBarSubmit,
-        searchTerm: props.searchTerm,
-      });
+      expect(wrapper.find(SearchBar));
```
Let's go with `expect(wrapper.find(SearchBar).exists()).toBeTruthy();` instead.

ttannis commented on amundsen_application/static/js/components/TableDetail/index.tsx:
```diff
@@ -323,14 +322,14 @@ export class TableDetail extends React.Component<TableDetailProps & RouteCompone
       innerContent = (
         <div className="container error-label">
-          <Breadcrumb path='/' text='Search Results'/>
+          <Breadcrumb path="/" text="Home" />
```
Was this a find/replace mistake? On the `TableDetail` page we want the logic of whether or not to go back to `Home` or `SearchPage` to be executed.

ttannis commented on amundsen_application/static/js/components/common/Breadcrumb/index.tsx:
```diff
@@ -1,14 +1,30 @@
+
+export type BreadcrumbProps = OwnProps & StateFromProps;
+
+export const Breadcrumb: React.SFC<BreadcrumbProps> = (props) => {
```
nit: If you do:
`export const Breadcrumb: React.SFC<BreadcrumbProps> = ({ path, text, searchTerm ) => {`
then variables `path` and `text` and `searchTerm` will automatically be defined and pulled from the `props`. Then, you can also set the default `path` and `text` in the `defaultProps` on line 40. This will make it such that all you have to have in this section is:
```
if (searchTerm) {
    path = `/search?searchTerm=${props.searchTerm}&selectedTab=table&pageIndex=0`
    text = 'Search Results'
}
```

ryanlieu replied:
:+1: yeah that's way cleaner

ryanlieu commented:
@ttannis just pushed that breadcrumb cleanup - going to 🚢after tests pass
